src/Nethermind/Nethermind.Blockchain/TransactionPools/TransactionPool.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                return;
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Added peer: {peer.ClientId}");
+            }
+        }
+
+        public void DeletePeer(NodeId nodeId)
+        {
+            if (!_peers.TryRemove(nodeId.PublicKey, out _))
+            {
+                return;
+            }
+
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Removed peer: {nodeId}");
+            }
+        }
+
+        public void AddTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            if (_pendingTransactions.ContainsKey(transaction.Hash))
+            {
+                return;
+            }
+
+            var recoveredAddress = _signer.RecoverAddress(transaction, blockNumber);
+            if (recoveredAddress != transaction.SenderAddress)
+            {
+                throw new InvalidOperationException("Invalid signature");
+            }
+
+            _pendingTransactions.TryAdd(transaction.Hash, transaction);
+            NewPending?.Invoke(this, new TransactionEventArgs(transaction));
+            NotifyPeers(SelectPeers(GetAvailablePeers(transaction)), transaction);
+            FilterAndStoreTransaction(transaction, blockNumber);
+        }
+
+        private void FilterAndStoreTransaction(Transaction transaction, UInt256 blockNumber)
+        {
+            var filters = _filters.Values.ToArray();
+            for (var i = 0; i < filters.Length; i++)
+            {
+                var filter = filters[i];
+                if (!filter.CanAdd(transaction))
+                {
+                    return;
+                }
+            }
+
+            _transactionStorage.Add(transaction, blockNumber);
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Added transaction: {transaction.Hash}");
+            }
+        }
+
+        public void DeleteTransaction(Keccak hash)
+        {
+            _pendingTransactions.TryRemove(hash, out var transaction);
+            var filters = _filters.Values.ToArray();
+            for (var i = 0; i < filters.Length; i++)
+            {
+                var filter = filters[i];
+                if (!filter.CanDelete(transaction))
+                {
+                    return;
+                }
+            }
+
+            _transactionStorage.Delete(hash);
+            if (_logger.IsDebug)
+            {
+                _logger.Debug($"Deleted transaction: {hash}");
+            }
+        }
+
+        public void AddReceipt(TransactionReceipt receipt)
+        {
+            _receiptStorage.Add(receipt);
+        }
+
+        public event EventHandler<TransactionEventArgs> NewPending;
+
+        private void NotifyPeers(ISynchronizationPeer[] peers, Transaction transaction)
+        {
+            if (peers.Length == 0)
+            {
+                return;
+            }
+
+            for (var i = 0; i < peers.Length; i++)
```

Comment:
Parallel.For?

Reply:
not sure if there is any waiting there, if not then this would probably be a waste